File: collector/raptor_collector/zone.py
"""BIND zone-file parser: $ORIGIN, trailing-dot names, SOA serial, common RR types."""

import re
from dataclasses import dataclass
from typing import Iterable, List, Optional

from raptor_collector.normalize import normalize_fqdn

_TTL_RE = re.compile(r"^(\d+)([smhdw])?$", re.IGNORECASE)
_CLASS_TOKENS = {"IN", "CH", "HS"}
_KEEP_TYPES = {"A", "AAAA", "CNAME", "TXT", "MX", "NS", "SOA"}


@dataclass(frozen=True)
class ResourceRecord:
    fqdn: str
    rrtype: str
    rdata: str
    ttl: Optional[int] = None
    zone: str = ""


def parse_ttl_token(token: str) -> Optional[int]:
    match = _TTL_RE.match(str(token or "").strip())
    if not match:
        return None
    value = int(match.group(1))
    unit = (match.group(2) or "").lower()
    multipliers = {"": 1, "s": 1, "m": 60, "h": 3600, "d": 86400, "w": 604800}
    return value * multipliers[unit]


def _strip_comment(line: str) -> str:
    in_quote = False
    chars: List[str] = []
    for char in line:
        if char == '"':
            in_quote = not in_quote
            chars.append(char)
            continue
        if char == ";" and not in_quote:
            break
        chars.append(char)
    return "".join(chars).rstrip()


def unfold_zone_text(text: str) -> List[str]:
    depth = 0
    buf: List[str] = []
    lines: List[str] = []
    for raw in str(text or "").splitlines():
        stripped = _strip_comment(raw)
        if not stripped.strip() and depth == 0:
            continue
        depth += stripped.count("(") - stripped.count(")")
        buf.append(stripped.replace("(", " ").replace(")", " "))
        if depth <= 0:
            joined = " ".join(part.strip() for part in buf if part.strip())
            if joined:
                lines.append(joined)
            buf = []
            depth = 0
    if buf:
        joined = " ".join(part.strip() for part in buf if part.strip())
        if joined:
            lines.append(joined)
    return lines


def _split_tokens(line: str) -> List[str]:
    return [token for token in re.findall(r'"[^"]*"|[^\s]+', line) if token]


def parse_zone_records(text: str, origin: str) -> List[ResourceRecord]:
    current_origin = normalize_fqdn(origin)
    default_ttl: Optional[int] = None
    last_owner = ""
    records: List[ResourceRecord] = []

    for line in unfold_zone_text(text):
        if line.upper().startswith("$ORIGIN"):
            parts = line.split(None, 1)
            if len(parts) == 2:
                current_origin = normalize_fqdn(parts[1], current_origin)
                last_owner = current_origin
            continue
        if line.upper().startswith("$TTL"):
            parts = line.split(None, 1)
            if len(parts) == 2:
                default_ttl = parse_ttl_token(parts[1].split()[0])
            continue
        if line.upper().startswith("$"):
            continue

        tokens = _split_tokens(line)
        if not tokens:
            continue

        owner = last_owner
        idx = 0
        ttl = default_ttl
        if tokens[0].upper() not in _CLASS_TOKENS and parse_ttl_token(tokens[0]) is None:
            maybe_type = tokens[0].upper().strip('"')
            if maybe_type not in _KEEP_TYPES:
                owner = normalize_fqdn(tokens[0], current_origin)
                idx = 1

        if idx < len(tokens):
            parsed_ttl = parse_ttl_token(tokens[idx])
            if parsed_ttl is not None:
                ttl = parsed_ttl
                idx += 1
        if idx < len(tokens) and tokens[idx].upper() in _CLASS_TOKENS:
            idx += 1
        if idx >= len(tokens):
            continue

        rrtype = tokens[idx].upper().strip('"')
        idx += 1
        if rrtype not in _KEEP_TYPES:
            last_owner = owner or last_owner
            continue

        rdata = " ".join(token.strip('"') for token in tokens[idx:]).strip()
        fqdn = owner or last_owner or current_origin
        if not fqdn or not rdata:
            last_owner = owner or last_owner
            continue
        records.append(
            ResourceRecord(fqdn=fqdn, rrtype=rrtype, rdata=rdata, ttl=ttl, zone=current_origin)
        )
        last_owner = fqdn
    return records


def soa_serial(records: Iterable[ResourceRecord]) -> Optional[str]:
    for record in records:
        if record.rrtype != "SOA":
            continue
        parts = str(record.rdata or "").split()
        if len(parts) >= 3 and parts[2].isdigit():
            return parts[2]
    return None


def records_to_payload(records: Iterable[ResourceRecord]) -> List[dict]:
    payload = []
    for record in records:
        item = {
            "fqdn": record.fqdn,
            "rrtype": record.rrtype,
            "rdata": record.rdata,
        }
        if record.ttl is not None:
            item["ttl"] = record.ttl
        payload.append(item)
    return payload

File: collector/internal/zone/zone.go
package zone

import (
	"regexp"
	"strconv"
	"strings"
)

type Record struct {
	FQDN   string
	RRType string
	RData  string
	TTL    int
	Zone   string
}

var ttlRe = regexp.MustCompile(`(?i)^(\d+)([smhdw])?$`)

var classTokens = map[string]bool{"IN": true, "CH": true, "HS": true}
var keepTypes = map[string]bool{"A": true, "AAAA": true, "CNAME": true, "TXT": true, "MX": true, "NS": true, "SOA": true}

func NormalizeFQDN(name, origin string) string {
	rawName := strings.TrimSpace(name)
	rawOrigin := strings.TrimSuffix(strings.ToLower(strings.TrimSpace(origin)), ".")
	if rawName == "" || rawName == "@" || rawName == "." {
		return rawOrigin
	}
	absolute := strings.HasSuffix(rawName, ".")
	normalized := strings.ToLower(strings.TrimSuffix(rawName, "."))
	if normalized == "" {
		return rawOrigin
	}
	if absolute || rawOrigin == "" {
		return normalized
	}
	if normalized == rawOrigin || strings.HasSuffix(normalized, "."+rawOrigin) {
		return normalized
	}
	return normalized + "." + rawOrigin
}

func parseTTL(token string) (int, bool) {
	m := ttlRe.FindStringSubmatch(strings.TrimSpace(token))
	if m == nil {
		return 0, false
	}
	value, _ := strconv.Atoi(m[1])
	mult := 1
	switch strings.ToLower(m[2]) {
	case "m":
		mult = 60
	case "h":
		mult = 3600
	case "d":
		mult = 86400
	case "w":
		mult = 604800
	}
	return value * mult, true
}

func stripComment(line string) string {
	inQuote := false
	var b strings.Builder
	for _, r := range line {
		if r == '"' {
			inQuote = !inQuote
			b.WriteRune(r)
			continue
		}
		if r == ';' && !inQuote {
			break
		}
		b.WriteRune(r)
	}
	return strings.TrimRight(b.String(), " \t")
}

func unfold(text string) []string {
	depth := 0
	var buf []string
	var lines []string
	for _, raw := range strings.Split(text, "\n") {
		stripped := stripComment(raw)
		if strings.TrimSpace(stripped) == "" && depth == 0 {
			continue
		}
		depth += strings.Count(stripped, "(") - strings.Count(stripped, ")")
		buf = append(buf, strings.ReplaceAll(strings.ReplaceAll(stripped, "(", " "), ")", " "))
		if depth <= 0 {
			joined := strings.Join(trimAll(buf), " ")
			if strings.TrimSpace(joined) != "" {
				lines = append(lines, strings.TrimSpace(joined))
			}
			buf = nil
			depth = 0
		}
	}
	if len(buf) > 0 {
		joined := strings.Join(trimAll(buf), " ")
		if strings.TrimSpace(joined) != "" {
			lines = append(lines, strings.TrimSpace(joined))
		}
	}
	return lines
}

func trimAll(parts []string) []string {
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if s := strings.TrimSpace(p); s != "" {
			out = append(out, s)
		}
	}
	return out
}

var tokenRe = regexp.MustCompile(`"[^"]*"|\S+`)

func Parse(text, origin string) []Record {
	current := NormalizeFQDN(origin, "")
	var defaultTTL int
	lastOwner := ""
	var records []Record
	for _, line := range unfold(text) {
		upper := strings.ToUpper(line)
		if strings.HasPrefix(upper, "$ORIGIN") {
			parts := strings.Fields(line)
			if len(parts) >= 2 {
				current = NormalizeFQDN(parts[1], current)
				lastOwner = current
			}
			continue
		}
		if strings.HasPrefix(upper, "$TTL") {
			parts := strings.Fields(line)
			if len(parts) >= 2 {
				if ttl, ok := parseTTL(parts[1]); ok {
					defaultTTL = ttl
				}
			}
			continue
		}
		if strings.HasPrefix(line, "$") {
			continue
		}
		tokens := tokenRe.FindAllString(line, -1)
		if len(tokens) == 0 {
			continue
		}
		owner := lastOwner
		idx := 0
		ttl := defaultTTL
		if !classTokens[strings.ToUpper(tokens[0])] {
			if _, isTTL := parseTTL(tokens[0]); !isTTL {
				maybe := strings.ToUpper(strings.Trim(tokens[0], `"`))
				if !keepTypes[maybe] {
					owner = NormalizeFQDN(tokens[0], current)
					idx = 1
				}
			}
		}
		if idx < len(tokens) {
			if parsed, ok := parseTTL(tokens[idx]); ok {
				ttl = parsed
				idx++
			}
		}
		if idx < len(tokens) && classTokens[strings.ToUpper(tokens[idx])] {
			idx++
		}
		if idx >= len(tokens) {
			continue
		}
		rrtype := strings.ToUpper(strings.Trim(tokens[idx], `"`))
		idx++
		if !keepTypes[rrtype] {
			if owner != "" {
				lastOwner = owner
			}
			continue
		}
		var rdataParts []string
		for _, tok := range tokens[idx:] {
			rdataParts = append(rdataParts, strings.Trim(tok, `"`))
		}
		rdata := strings.TrimSpace(strings.Join(rdataParts, " "))
		fqdn := owner
		if fqdn == "" {
			fqdn = lastOwner
		}
		if fqdn == "" {
			fqdn = current
		}
		if fqdn == "" || rdata == "" {
			continue
		}
		records = append(records, Record{FQDN: fqdn, RRType: rrtype, RData: rdata, TTL: ttl, Zone: current})
		lastOwner = fqdn
	}
	return records
}

func SOASerial(records []Record) string {
	for _, rec := range records {
		if rec.RRType != "SOA" {
			continue
		}
		parts := strings.Fields(rec.RData)
		if len(parts) >= 3 {
			if _, err := strconv.Atoi(parts[2]); err == nil {
				return parts[2]
			}
		}
	}
	return ""
}

func ToPayload(records []Record) []map[string]any {
	out := make([]map[string]any, 0, len(records))
	for _, rec := range records {
		item := map[string]any{"fqdn": rec.FQDN, "rrtype": rec.RRType, "rdata": rec.RData}
		if rec.TTL > 0 {
			item["ttl"] = rec.TTL
		}
		out = append(out, item)
	}
	return out
}
